tests/unit/test_kernel.py

```diff
@@ -1681,8 +2079,46 @@ def k_x_y(x_input, y_input):
 
     # pylint: enable=too-many-locals
 
+    def test_stein_kernel_invalid_base_kernel(self) -> None:
+        r"""
+        Test how the SteinKernel handles an invalid kernel being passed.
+        """
+        # Setup some data
+        num_points_x = 10
+        num_points_y = 5
+        dimension = 2
+
+        def score_function(x_: ArrayLike) -> Array:
+            """
+            Compute a simple, example score function for testing purposes.
+
+            :param x_: Point or points at which we wish to evaluate the score function
+            :return: Evaluation of the score function at ``x_``
+            """
+            return -x_
+
+        # Setup data
+        generator = np.random.default_rng(1_989)
+        x = generator.random((num_points_x, dimension))
+        y = generator.random((num_points_y, dimension))
+
+        # Compute output using Kernel class - since the base kernel does not have a
+        # compute method, and it does not have a defined pytree, it should give an error
+        # from JAX internally, reference the specific class that has been passed and the
+        # compute line that fails, which is enough information for the user to debug
+        # with.
+        kernel = coreax.kernel.SteinKernel(
+            base_kernel=coreax.util.InvalidKernel(x=1.0),
+            score_function=score_function,
+        )
+        with self.assertRaises(TypeError) as error_raised:
+            kernel.compute(x, y)
 
-# pylint: enable=too-many-public-methods
+        self.assertEqual(
+            error_raised.exception.args[0],
+            "Cannot interpret value of type <class 'coreax.util.InvalidKernel'> "
+            "as an abstract array; it does not have a dtype attribute",
```

Comment:
This isn't a helpful error message - the issue isn't that `InvalidKernel` doesn't have a `dtype` attribute (nor do any of the `Kernel` subclasses!) but rather that it's not a `Kernel`.

If you naively follow the error message and add a `dtype` to your custom kernel, you get a different error message:

``` python
class MyKernel:
    dtype = int

    def __init__(self, x: float):
        self.x = x

...

kernel = coreax.kernel.SteinKernel(
    base_kernel=MyKernel(x=1.0),
    score_function=score_function,
)

kernel.compute(x, y)
```

you get a different, slightly more relevant-looking error: `AttributeError: DynamicJaxprTracer has no attribute compute_elementwise`

But this is also not the correct solution, as if we update our custom kernel to have a `compute_elementwise`

``` python
class MyKernel:
    dtype = int

    def __init__(self, x: float):
        self.x = x

    def compute_elementwise(self, x: ArrayLike, y: ArrayLike) -> Array:
        return x + y
```

we get the exact same error (`AttributeError: DynamicJaxprTracer has no attribute compute_elementwise`). 

In order for this to be useful to the end user, we should probably just do one of the following:

- In `SteinKernel.compute_elementwise`, catch any TypeError or AttributeError and check that `base_kernel` is a `Kernel` instance; if not, re-raise the error as a TypeError pointing out that it should be a `Kernel` instance  
- In `SteinKernel.__init__`, check that `base_kernel` really is a `Kernel` instance, and throw a TypeError if not

Reply:
I agree that this error message is not helpful and needs updating. However, I am a bit unsure as to why the DynamicJaxprTracer cannot see the compute_elementwise method on this example (invalid) kernel. @tc85324 - do you have any idea why, when passing the `SteinKernel` this example `MyKernel` object in this chunk of code here, that it would have an issue using `compute_elementwise` when it is actually defined on the class? 

As a side note, I think we wanted to move away from asserting things were specific objects if possible (another object may be passed, as long as it works) and I am not sure if specifying things in the init of the Stein kernel will cause problems when JAX integration occurs.

Reply:
I suppose as a workaround we could just define a small `@runtime_checkable` `Protocol` that does the same thing if we don't want to assert that things inherit from specific classes, just that the relevant methods are present?

Reply:
@rg936672  I've just pushed some changes - the issue was that JAX requires methods that inform it how to flatten and build pytrees for provided classes, or it errors internally. I've now added a check that the base kernel has the required methods, with helpful error messages if not. I've also expanded the tests to cover a case where it can build the pytree, but is still missing a computational method, which errors with a sensible text. Are you happy with this?

Reply:
Looks good! For completeness' sake we'd ideally have a test for each of the required methods, checking the error message is appropriate, but I appreciate that'd result in a lot of boilerplate (especially without Pytest) so not urgent.